NXP imx95 add basic pm support

yongxu-wang15 opened the pull request:
added basic pm support for imx95 evk m7 board, verified it in power_mgmt_soc test case, based on this case, M7 can enter/exit wait/suspend mode as we expected.

yongxu-wang15 committed:
dts: arm: nxp_imx95_m7: add pm node

added wait/stop/suspend pm node for imx95 m7

Signed-off-by: Yongxu Wang <yongxu.wang@nxp.com>

yongxu-wang15 committed:
soc: nxp: imx9: add basic pm process for i.MX95 M7

add basic pm_state_set and pm_state_exit_post_ops

Signed-off-by: Yongxu Wang <yongxu.wang@nxp.com>

yongxu-wang15 committed:
pm: power_mgmt_soc: add imx95_evk m7 into test case yml

enable imx95_evk m7 for this power_mgmt_soc test case

Signed-off-by: Yongxu Wang <yongxu.wang@nxp.com>

yongxu-wang15 committed:
drivers: scmi: add use_polling parameter in scmi_send_message

SCMI supports both polling and interrupt modes for message completion.
Previously, the scmi_send_message() API used a 'pre_kernel' flag to
determine which mode to use_polling during the pre-kernel phase and
interrupts post-kernel.

This approach tightly coupled the decision logic with kernel state,
limiting flexibility. In particular, certain power management (PM)
related SCMI APIs require polling mode even in post-kernel context
to avoid unintended CPU wakeups caused by interrupts.

This patch replaces the 'pre_kernel' with a more generic
'use_polling' parameter, allowing callers to explicitly specify
the desired behavior. Typical usage can still rely on k_is_pre_kernel()
to determine polling mode in higher level api, while PM related
calls can directly enforce polling regardless of kernel state.

Signed-off-by: Yongxu Wang <yongxu.wang@nxp.com>

bjarki-andreasen reviewed (changes requested):
Don't know anything about SCMI, but at least some irq_disable() may be superfluous :)

PetervdPerk-NXP reviewed (commented):
Could you verify if the systick is actually capable of waking up the core when doing wfi?
We've seen some issues with this, that other interrupts can wake the m7 but the systisck doesn't thus giving big delays.

yongxu-wang15 commented on drivers/firmware/scmi/core.c:
```diff
@@ -169,7 +176,16 @@ int scmi_send_message(struct scmi_protocol *proto, struct scmi_message *msg,
+	 * the sm interrupt should be disabled in the soc low power phase.
+	 * In this case, the pre kernel is used instead of the post kernel.
+	 */
+	use_pre_kernel_path = use_pre_kernel_path || scmi_is_current_thread_idle();
```
thanks, updated

yongxu-wang15 commented on soc/nxp/imx/imx9/imx95/m7/soc.c:
```diff
@@ -88,6 +88,67 @@ static int soc_init(void)
+	irq_unlock(0);
+
+	/* disabled system manager irq */
+	irq_disable(DT_IRQN(DT_NODELABEL(mu5)));
```
the above code only mask irq to don't give reaction to interrupt, the capacity of wakeup Mcore from wfi is reserved, and the sm protocol use interrupt from scmi notification logic even though use scmi pre_kernel logic, so we need disabled mu5 now.

PetervdPerk-NXP replied:
I agree @JiafeiPan i still don't get why mu5 irq needs to be disabled if IRQ's are disabled already.

yongxu-wang15 replied:
1. the system manager will reply message in currently mechanism, so we need keep mu5 is be disabled to not wakeup m7 core from wfi after run scmi interface api.
2. the _disable_irq is to set PRIMASK register, it  will mask all regular peripheral interrupt responce for cpu(such as m7 core), it mean that m7 can't enter nvic irq handler in this side, but this will not shield the triggering of interrupts, which will cause the CPU to be awakened from WFI. So in low power use case, although we have already set _disable_irq, but the cpu still have recognize this MU interrupt and thus leave WFI.

PetervdPerk-NXP replied:
Isn't then the system design of system manager flawed? And needs reconsideration?

I think Linux uses the same mechanism using SCMI for PM does that work well then?

PetervdPerk-NXP replied:
Yes putting SCMI into non-interrupt mode for PM makes sense. Then poll response after wake up, this doesn't do busy waiting. Then after wakeup put SCMI MU into interrupt mode.

yongxu-wang15 replied:
sure, have updated it, now in pre kernel process(include idle task pm case), it will disable shem completion interrupt and mu side interrupt, and resume it after send message done. please help review it again

yongxu-wang15 commented on soc/nxp/imx/imx9/imx95/m7/soc.c:
```diff
@@ -88,6 +88,67 @@ static int soc_init(void)
+		scmi_cpu_sleep_mode_set(&cpu_cfg);
+		__DSB();
+		__WFI();
+		__ISB();
```
updated. thanks

yongxu-wang15 commented on soc/nxp/imx/imx9/imx95/m7/soc.c:
```diff
@@ -88,6 +88,67 @@ static int soc_init(void)
+	 */
+	__disable_irq();
+	/* Set BASEPRI to 0 */
+	irq_unlock(0);
```
hi @ceolin  @bjarki-andreasen , this is as imx95 m7 is based on ARMv7-M architecture, For this architecture, the current implementation of arch_irq_lock of zephyr is based on BASEPRI, which will only retain abnormal interrupts such as NMI,
 and all other interrupts from the CPU(including systemtick) will be masked, which makes the CORE unable to be woken up from WFI. so we used primask rather than defaultly basepri in there

yongxu-wang15 replied:
if we want keep arch_irq_lock keep wakeup source capacity for all arch, it have some issue from my understand, https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/arch/arm/cortex_m/exception.h#L53
![image](https://github.com/user-attachments/assets/b1b18ee2-c45d-4dc8-b6a1-8f152f6c195f)

yongxu-wang15 commented on dts/arm/nxp/nxp_imx95_m7.dtsi:
```diff
@@ -27,6 +28,37 @@
+		};
+
+		power-states {
+			active: power-state-active {
```
thanks reminder, have deleted it

yongxu-wang15 commented:
hi @PetervdPerk-NXP , in my side, the system tick work fine on this patch

bjarki-andreasen reviewed (changes requested):
Regarding __disable_irq(), its soc dependent, so if it works, its fine. Regarding using internal kernel apis to check for idle thread context, that seems like a clear layering violation, and I have proposed a solution which seems to be more appropriate :)

yongxu-wang15 committed:
drivers: firmware: scmi: disable interrupt communication in polling mode

SCMI processor interrupts from agent are enabled by default during the
`setup_chan` initialization phase. This is suitable for interrupt-driven
communication, where the agent triggers an interrupt to notify the platform
after completing a message transaction.

However, when using the polling model, interrupts are not required and may
cause unintended behavior or performance issues in PM.
To ensure correct operation under polling mode, the interrupt should
be explicitly disabled after channel setup.

Signed-off-by: Yongxu Wang <yongxu.wang@nxp.com>

yongxu-wang15 commented:
update this patch to re-name cpu into cpu0: cpu@0 in order to align pm test case, please help review this pr, thanks a lot

yongxu-wang15 committed:
drivers: firmware: scmi: add cpu lpm interface api

Added scmi_cpu_pd_lpm_set api for nxp imx scmi interface

This api set the lpm setting for some peripherals applied
when cpu enter a low power mode, such as keep iMX95 wakeup mix
power on when M7 core enter suspend mode, scmi agent record and
deal with this request

Signed-off-by: Yongxu Wang <yongxu.wang@nxp.com>

yongxu-wang15 committed:
soc: nxp: imx95: setup m7mix and wakeupmix power on in suspend mode

wakeupmix keep power on state is essential for system
suspend mode, because of console uart locate in it.

temporarily set the M7 mix to power on,
further optimization will be carried out later

Signed-off-by: Yongxu Wang <yongxu.wang@nxp.com>

yongxu-wang15 commented on drivers/firmware/scmi/core.c:
```diff
@@ -169,7 +170,15 @@ int scmi_send_message(struct scmi_protocol *proto, struct scmi_message *msg,
+	 * the sm interrupt should be disabled in the soc low power phase.
+	 * In this case, the pre kernel is used instead of the post kernel.
+	 */
+	use_pre_kernel_path |= z_is_idle_thread_object(_current);
```
hi @bjarki-andreasen , thanks your suggestion, have updated this flow, add pre_kernel variable into scmi_send_message api, and do kernel state judge in scmi lower level api

yongxu-wang15 commented:
newly update: 
1. from @bjarki-andreasen suggestion, add pre_kernel variable into scmi_send_message api, and do kernel state judge in scmi lower level api
2. add scmi nxp cpu lpm interface, and set M7/wakeup mix to keep power on state when m7 cpu enter suspend mode

bjarki-andreasen reviewed (dismissed):
Nice work! looks clean :)

bjarki-andreasen commented:
ping @ceolin

yongxu-wang15 commented:
newly update:  rebase code base to fix tests/subsys/pm/power_mgmt_soc/testcase.yaml conflict, please help review it when you free, thanks a lot

dbaluta commented on drivers/firmware/scmi/nxp/cpu.c:
```diff
@@ -45,3 +45,39 @@ int scmi_cpu_sleep_mode_set(struct scmi_cpu_sleep_mode_config *cfg)
+
+int scmi_cpu_pd_lpm_set(struct scmi_cpu_pd_lpm_config *cfg)
+{
+	struct scmi_protocol *proto = &SCMI_PROTOCOL_NAME(SCMI_PROTOCOL_CPU_DOMAIN);
```
What is LPM please at list reference this with the full name in the commit message.

yongxu-wang15 commented on drivers/firmware/scmi/clk.c:
```diff
@@ -6,6 +6,7 @@
 
```
thanks review, update it in newly commit, thanks

yongxu-wang15 commented on include/zephyr/drivers/firmware/scmi/nxp/cpu.h:
```diff
@@ -33,6 +35,23 @@ struct scmi_cpu_sleep_mode_config {
+	uint32_t lpmSetting;
+	uint32_t retMask;
+};
+
```
sure, done

yongxu-wang15 commented on drivers/firmware/scmi/core.c:
```diff
@@ -87,12 +88,34 @@ static int scmi_core_setup_chan(const struct device *transport,
+
+	return mbox_set_enabled_dt(tx_reply, enable);
+}
+
```
add more description about this api, thanks

yongxu-wang15 commented:
newly version update according into @dbaluta review
1. add use_polling parameter in scmi_send_message api, put selection about scmi mode between poll with interrupt into higher level scmi api, and in generic interface also use pre_kernel to judge, but directly set poll about PM.
2. update some inaccurate description of the problem
thanks

dbaluta commented on include/zephyr/drivers/firmware/scmi/protocol.h:
```diff
@@ -115,8 +115,10 @@ int scmi_status_to_errno(int scmi_status);
  * @retval 0 if successful
  * @retval negative errno if failure
+ * @param use_polling if use poll mode in currently scmi interface context
  */
```
use_polling:
      - possbile values: true, uses polling -> explain here what pollings is about
      - false, uses interrupt mode, explain here what interrupts mode means. (brief explanation).

yongxu-wang15 replied:
sure, done, thanks

yongxu-wang15 commented:
Updated: to resort pr by comments from @dbaluta
thanks

yongxu-wang15 commented on drivers/firmware/scmi/core.c:
```diff
@@ -87,7 +87,7 @@ static int scmi_core_setup_chan(const struct device *transport,
 
-static int scmi_send_message_pre_kernel(struct scmi_protocol *proto,
+static int scmi_send_message_polling(struct scmi_protocol *proto,
 					struct scmi_message *msg,
```
sure, added description for poll api, done

yongxu-wang15 commented on drivers/firmware/scmi/core.c:
```diff
@@ -113,6 +144,11 @@ static int scmi_send_message_polling(struct scmi_protocol *proto,
 
+	/* restore scmi interrupt enable status when disable it pass */
+	if (ret_disable_irq_status >= 0) {
+		scmi_interrupt_enable(proto->tx, true);
```
thanks to point out this issue, done

dbaluta commented:
You can have my approval (Reviewed-by) for SCMI API changes.  For PM changes we need others to review. 

Maybe in the future for such changes you can create 2 PRs. One that introduces API changes and can be merged faster and another one that is implementing an actual feature separated.

yongxu-wang15 commented:
hi @bjarki-andreasen  Could you please help to review again the PR when you get any chance?
thanks a lot

yongxu-wang15 commented:
Hi @ceolin , When you helped review it before, you left a change request. Could you please review it again if you have some free time? thanks
